Map subscript operator fails with keys of complex type

```
select map(array[array[1,1]], array['a'])[array[1,1]]
```

fails with:

```
Query 20160621_003752_01271_b4cxm failed: expected (Block,Block)boolean but found (Object,Object)boolean
com.facebook.presto.spi.PrestoException: expected (Block,Block)boolean but found (Object,Object)boolean
	at com.facebook.presto.operator.scalar.MapSubscriptOperator.subscript(MapSubscriptOperator.java:177)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:649)
	at com.facebook.presto.sql.planner.ExpressionInterpreter.invoke(ExpressionInterpreter.java:1093)
	at com.facebook.presto.sql.planner.ExpressionInterpreter$Visitor.invokeOperator(ExpressionInterpreter.java:1049)
	at com.facebook.presto.sql.planner.ExpressionInterpreter$Visitor.visitSubscriptExpression(ExpressionInterpreter.java:1016)
	at com.facebook.presto.sql.tree.SubscriptExpression.accept(SubscriptExpression.java:47)
	at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:22)
	at com.facebook.presto.sql.planner.ExpressionInterpreter.optimize(ExpressionInterpreter.java:231)
	at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.simplifyExpression(SimplifyExpressions.java:153)
	at com.google.common.collect.Maps$7.transformEntry(Maps.java:1812)
	at com.google.common.collect.Maps$10.getValue(Maps.java:1857)
	at com.google.common.collect.ImmutableMap.copyOf(ImmutableMap.java:292)
	at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions.java:110)
	at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions.java:88)
	at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:76)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:86)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:97)
	at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions.java:109)
	at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions.java:88)
	at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:76)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:86)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:97)
	at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions.java:109)
	at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions.java:88)
	at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:76)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:86)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:97)
	at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions.java:109)
	at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions.java:88)
	at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:76)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:86)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:75)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:76)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:40)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:24)
	at com.facebook.presto.sql.planner.plan.PlanVisitor.visitOutput(PlanVisitor.java:50)
	at com.facebook.presto.sql.planner.plan.OutputNode.accept(OutputNode.java:81)
	at com.facebook.presto.sql.planner.plan.SimplePlanRewriter.rewriteWith(SimplePlanRewriter.java:29)
	at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions.optimize(SimplifyExpressions.java:85)
	at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:100)
	at com.facebook.presto.execution.SqlQueryExecution.doAnalyzeQuery(SqlQueryExecution.java:277)
	at com.facebook.presto.execution.SqlQueryExecution.analyzeQuery(SqlQueryExecution.java:256)
	at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:220)
	at com.facebook.presto.execution.QueuedExecution.lambda$start$1(QueuedExecution.java:68)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.invoke.WrongMethodTypeException: expected (Block,Block)boolean but found (Object,Object)boolean
	at java.lang.invoke.Invokers.newWrongMethodTypeException(Invokers.java:298)
	at java.lang.invoke.Invokers.checkExactType(Invokers.java:309)
	at com.facebook.presto.operator.scalar.MapSubscriptOperator.subscript(MapSubscriptOperator.java:170)
	... 55 more
```

The following call in `MapSubscriptOperator.subscript()` for the default (Object) specialization should probably needs to be changed to use `invoke` instead of `invokeExact`:

```java
if ((boolean) keyEqualsMethod.invokeExact((keyType.getObject(map, position), key)) {
```

--- a/presto-main/src/main/java/com/facebook/presto/operator/scalar/MapSubscriptOperator.java
+++ b/presto-main/src/main/java/com/facebook/presto/operator/scalar/MapSubscriptOperator.java
@@ -172,7 +172,7 @@ public class MapSubscriptOperator
     {
         for (int position = 0; position < map.getPositionCount(); position += 2) {
             try {
-                if ((boolean) keyEqualsMethod.invokeExact(keyType.getObject(map, position), key)) {
+                if ((boolean) keyEqualsMethod.invoke(keyType.getObject(map, position), key)) {
                     return readNativeValue(valueType, map, position + 1); // position + 1: value position
                 }
             }

--- a/presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java
+++ b/presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java
@@ -1174,6 +1174,7 @@ public class TestExpressionInterpreter
         optimize("MAP(ARRAY [1, 2], ARRAY [3, 4])[0]");
         optimize("MAP(ARRAY [BIGINT '1', 2], ARRAY [3, 4])[0]");
         optimize("MAP(ARRAY [1, 2], ARRAY [3, 4])[5]");
+        optimize("MAP(ARRAY [ARRAY[1,1]], ARRAY['a'])[ARRAY[1,1]]");
     }
 
     @Test(timeOut = 60000)

--- a/presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java
+++ b/presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java
@@ -190,6 +190,16 @@ public abstract class AbstractTestQueries
     }
 
     @Test
+    public void testMapSubscript()
+            throws Exception
+    {
+        assertQuery("select map(array[1], array['aa'])[1]", "select 'aa'");
+        assertQuery("select map(array['a'], array['aa'])['a']", "select 'aa'");
+        assertQuery("select map(array[array[1,1]], array['a'])[array[1,1]]", "select 'a'");
+        assertQuery("select map(array[(1,2)], array['a'])[(1,2)]", "select 'a'");
+    }
+
+    @Test
     public void testVarbinary()
             throws Exception
     {

